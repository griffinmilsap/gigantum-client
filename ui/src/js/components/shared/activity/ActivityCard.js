// vendor
import React, { Component } from 'react';
import classNames from 'classnames';
// components
import ActivityDetails from 'Components/shared/activity/ActivityDetails';
// assets
import './ActivityCard.scss';

export default class ActivityCard extends Component {
  constructor(props) {
    super(props);

    const showDetails = props.edge.node.show && props.edge.node.detailObjects.filter(details => details.show).length !== 0;

    this.state = {
      showExtraInfo: showDetails,
      show: true,
      isOver: false,
    };
  }

  /**
    @param {string} timestamp
    if input is undefined. current time of day is used
    inputs a time stamp and return the time of day HH:MM am/pm
    @return {string}
  */
  _getTimeOfDay(timestamp) {
    const time = (timestamp !== undefined) ? new Date(timestamp) : new Date();
    const hour = (time.getHours() % 12 === 0) ? 12 : time.getHours() % 12;
    const minutes = (time.getMinutes() > 9) ? time.getMinutes() : `0${time.getMinutes()}`;
    const ampm = time.getHours() >= 12 ? 'pm' : 'am';

    return `${hour}:${minutes}${ampm}`;
  }

  /**
    @param {}
    * determines the appropriate margins for cards when compressing
    @return {object}
  */
  _processCardStyle() {
    const activityCardStyle = {};


    const { props, state } = this;

    if (props.isCompressed) {
      if (!props.isExpandedHead && !props.isExpandedEnd) {
        activityCardStyle.marginTop = '-7.5px';
        activityCardStyle.marginBottom = '-7.5px';
      } else if (props.isExpandedHead) {
        const distance = props.attachedCluster.length - 1;
        const calculatedMargin = distance * 7.5;

        activityCardStyle.marginTop = `${calculatedMargin}px`;
        activityCardStyle.marginBottom = '-7.5px';
      } else if (props.isExpandedEnd) {
        const distance = props.attachedCluster.length - 1;
        const calculatedMargin = distance * 7.5;

        activityCardStyle.marginTop = '-7.5px';
        activityCardStyle.marginBottom = `${calculatedMargin}px`;
      }
    }

    return activityCardStyle;
  }


  _expandCollapseSideBar(isOver) {
    const { props } = this;
    props.compressExpanded(props.attachedCluster, !isOver);

    this.setState({ isOver });
  }


  render() {
    const { props, state } = this;
    const node = props.edge.node;


    const type = node.type && node.type.toLowerCase();

    const activityCardCSS = classNames({
      'ActivityCard card': state.showExtraInfo,
      'ActivityCard ActivityCard--collapsed card': !state.showExtraInfo,
      'ActivityCard--faded': props.hoveredRollback > props.position,
      ActivityCard__expanded: props.isExpandedNode,
      ActivityCard__compressed: props.isCompressed && !props.isExpandedEnd,
    });


    const titleCSS = classNames({
      'ActivityCard__title flex flex--row justify--space-between': true,
      open: state.showExtraInfo || (type === 'note' && state.show),
      closed: !state.showExtraInfo || (type === 'note' && !state.show),
    });


    const expandedCSS = classNames({
      ActivityCard__node: props.isExpandedNode && !props.isExpandedHead && !props.isExpandedEnd,
      'ActivityCard__start-node': props.isExpandedHead,
      'ActivityCard__end-node': props.isExpandedEnd,
    });

    const expandedHeight = props.attachedCluster ? 110 * (props.attachedCluster.length - 1) : 0;

    const expandedBarHeight = state.isOver ? expandedHeight - ((props.attachedCluster.length - 1) * 15) : expandedHeight;

    const margin = state.isOver ? ((props.attachedCluster.length - 1) * 7.5) : 0;

    const expandedContainerStyle = props.attachedCluster && {
      height: `${expandedHeight}px`,
    };

    const expandedStyle = props.attachedCluster && {
      height: `${expandedBarHeight}px`,
      margin: `${margin}px 0 0 0`,
    };

    if (props.isFirstCard && props.attachedCluster) {
      expandedStyle.top = '2px';
    }

    const activityCardStyle = this._processCardStyle();
    return (
      <div className="ActivityCard__container column-1-span-9">

        <div className={activityCardCSS} ref="card" style={activityCardStyle}>

          <div
            className={`ActivityCard__badge ActivityCard__badge--${type}`}
            title={type}
          />

          <div className="ActivityCard__content">

            <div
              className={titleCSS}
              onClick={() => this.setState({ show: !state.show, showExtraInfo: !state.showExtraInfo })}
            >

              <div className="ActivityCard__stack">

                <p className="ActivityCard__time">
                  {this._getTimeOfDay(props.edge.node.timestamp)}
                </p>

                <div className="ActivityCard__user" />
              </div>

              <h6 className="ActivityCard__commit-message">
                <b>{`${props.edge.node.username} - `}</b>
                { props.edge.node.message }
              </h6>

            </div>

            {
              (state.showExtraInfo && (type !== 'note' || state.show))
                && (
                <ActivityDetails
                  sectionType={props.sectionType}
                  edge={props.edge}
                  show={state.showExtraInfo}
                  key={`${node.id}_activity-details`}
                  node={node}
                />
                )
            }
          </div>
        </div>
      </div>
    );
  }
}
