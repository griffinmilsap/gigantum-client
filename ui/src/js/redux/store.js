import {
  createStore,
  compose,
  applyMiddleware,
} from 'redux';
import reducers from './combinedReducers';
import { persistPreferences } from '../utils/localStorage';
import { getPreferences } from './preferences/preferences';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const localStorageMiddleware = store => next => (action) => {
  const result = next(action);

  persistPreferences(getPreferences(store.getState()));

  return result;
};

export default createStore(
  reducers,
  composeEnhancers(applyMiddleware(localStorageMiddleware)),
);
