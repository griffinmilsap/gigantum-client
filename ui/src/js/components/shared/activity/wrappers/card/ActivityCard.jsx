// @flow
// vendor
import React, { Component } from 'react';
import classNames from 'classnames';
// components
import ActivityDetails from './ActivityDetails';
// assets
import './ActivityCard.scss';

/**
  @param {Object} props
  logic if the card should show by default or not
  @return {Boolean}
*/
const shouldShowDetails = (props) => {
  const { node } = props.edge;
  const { show, detailObjects } = node;
  const detailsHasShow = detailObjects.filter(details => details.show).length !== 0;

  return (show && detailsHasShow);
};

/**
  @param {Object} props
  @param {Object} state
  gets varibales for render function
  @return {Object}
*/
const getRenderVariables = (props, state) => {
  const { node } = props.edge;
  const type = node.type && node.type.toLowerCase();
  const expandedHeight = props.attachedCluster
    ? 110 * (props.attachedCluster.length - 1)
    : 0;
  const expandedBarHeight = state.isOver
    ? expandedHeight - ((props.attachedCluster.length - 1) * 15)
    : expandedHeight;
  const margin = state.isOver
    ? ((props.attachedCluster.length - 1) * 7.5)
    : 0;

  // declare style here
  const expandedStyle = props.attachedCluster && {
    height: `${expandedBarHeight}px`,
    margin: `${margin}px 0 0 0`,
  };
  const firstInitial = props.edge.node.username.charAt(0).toUpperCase();

  if (props.isFirstCard && props.attachedCluster) {
    expandedStyle.top = '2px';
  }

  return {
    expandedBarHeight,
    expandedHeight,
    expandedStyle,
    firstInitial,
    margin,
    type,
  };
};


type Props = {
  attachedCluster: Array,
  compressExpanded: Function,
  edge: {
    node: {
      message: String,
      username: String,
    },
  },
  hoveredRollback: Boolean,
  isCompressed: Boolean,
  isExpandedHead: Boolean,
  isExpandedEnd: Boolean,
  isExpandedNode: Boolean,
  name: string,
  owner: string,
  position: Number,
  sectionType: String,
};

class ActivityCard extends Component<Props> {
  state = {
    showExtraInfo: shouldShowDetails(this.props),
    show: true,
    isOver: false,
  }

  /**
    @param {string} timestamp
    if input is undefined. current time of day is used
    inputs a time stamp and return the time of day HH:MM am/pm
    @return {string}
  */
  _getTimeOfDay = (timestamp) => {
    const time = (timestamp !== undefined) ? new Date(timestamp) : new Date();
    const hour = (time.getHours() % 12 === 0) ? 12 : time.getHours() % 12;
    const minutes = (time.getMinutes() > 9) ? time.getMinutes() : `0${time.getMinutes()}`;
    const ampm = time.getHours() >= 12 ? 'pm' : 'am';

    return `${hour}:${minutes}${ampm}`;
  }

  /**
    @param {}
    * determines the appropriate margins for cards when compressing
    @return {object}
  */
  _processCardStyle = () => {
    const activityCardStyle = {};
    const {
      attachedCluster,
      isCompressed,
      isExpandedHead,
      isExpandedEnd,
    } = this.props;

    if (isCompressed) {
      if (!isExpandedHead && !isExpandedEnd) {
        activityCardStyle.marginTop = '-7.5px';
        activityCardStyle.marginBottom = '-7.5px';
      } else if (isExpandedHead) {
        const distance = attachedCluster.length - 1;
        const calculatedMargin = distance * 7.5;

        activityCardStyle.marginTop = `${calculatedMargin}px`;
        activityCardStyle.marginBottom = '-7.5px';
      } else if (isExpandedEnd) {
        const distance = attachedCluster.length - 1;
        const calculatedMargin = distance * 7.5;

        activityCardStyle.marginTop = '-7.5px';
        activityCardStyle.marginBottom = `${calculatedMargin}px`;
      }
    }

    return activityCardStyle;
  }

  /**
    @param {Boolean} isOver
    sets state for side bar to compress on hover
  */
  _expandCollapseSideBar = (isOver) => {
    const { attachedCluster, compressExpanded } = this.props;
    compressExpanded(attachedCluster, !isOver);

    this.setState({ isOver });
  }

  /**
    @param {} isOver
    toggles show and showExtraInfo
  */
  _toggleDetails = () => {
    this.setState((state) => {
      const show = !state.show;
      const showExtraInfo = !state.showExtraInfo;
      return { show, showExtraInfo };
    });
  }

  render() {
    const {
      show,
      showExtraInfo,
    } = this.state;
    const {
      edge,
      hoveredRollback,
      isCompressed,
      isExpandedEnd,
      isExpandedNode,
      name,
      owner,
      position,
      sectionType,
    } = this.props;
    const { node } = edge;
    const activityCardStyle = this._processCardStyle();
    const {
      firstInitial,
      type,
    } = getRenderVariables(this.props, this.state);
    // declare css here
    const activityCardCSS = classNames({
      'ActivityCard card': showExtraInfo,
      'ActivityCard ActivityCard--collapsed card': !showExtraInfo,
      'ActivityCard--faded': hoveredRollback > position,
      ActivityCard__expanded: isExpandedNode,
      ActivityCard__compressed: isCompressed && !isExpandedEnd,
    });
    const titleCSS = classNames({
      'ActivityCard__title flex flex--row justify--space-between': true,
      open: showExtraInfo || (type === 'note' && show),
      closed: !showExtraInfo || (type === 'note' && !show),
    });

    return (
      <div className="ActivityCard__container column-1-span-9">
        <div className={activityCardCSS} style={activityCardStyle}>

          <div
            className={`ActivityCard__badge ActivityCard__badge--${type}`}
            title={type}
          />

          <div className="ActivityCard__content">

            <div
              className={titleCSS}
              onClick={() => this._toggleDetails()}
              role="presentation"
            >
              <div className="ActivityCard__stack">

                <p className="ActivityCard__time">
                  {this._getTimeOfDay(edge.node.timestamp)}
                </p>

                <div className="ActivityCard__user">{firstInitial}</div>
              </div>

              <h6 className="ActivityCard__commit-message">
                <b>{`${edge.node.username} - `}</b>
                { edge.node.message }
              </h6>
            </div>
            {
              (showExtraInfo && ((type !== 'note') || show))
                && (
                  <ActivityDetails
                    sectionType={sectionType}
                    edge={edge}
                    show={showExtraInfo}
                    key={`${node.id}_activity-details`}
                    node={node}
                    owner={owner}
                    name={name}
                  />
                )
            }
          </div>
        </div>
      </div>
    );
  }
}


export default ActivityCard;
