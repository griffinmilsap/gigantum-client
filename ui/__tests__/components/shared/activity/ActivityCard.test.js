// vendor
import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import relayTestingUtils from '@gigantum/relay-testing-utils';
import { Provider } from 'react-redux';
// components
import ActivityCard from 'Components/shared/activity/wrappers/card/ActivityCard';
// Data
import json from 'Tests/components/labbook/__relaydata__/LabbookContainerQuery.json';
// store
import store from 'JS/redux/store';

let { labbook } = json.data;

let fixtures = {
  sectionType: 'labbook',
  isFirstCard: true,
  addCluster: jest.fn(),
  compressExpanded: jest.fn(),
  isCompressed: false,
  isExpandedHead: true,
  isExpandedEnd: true,
  isExpandedNode: true,
  attachedCluster: false,
  collapsed: false,
  clusterObject: {},
  position: 0,
  hoveredRollback: true,
  key: 'ActivityCard',
  edge: labbook.activityRecords.edges[0],
};


describe('Activity', () => {
  it('renders a snapshot', () => {

    const wrapper = renderer.create(<ActivityCard {...fixtures} />);

    const tree = wrapper.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
