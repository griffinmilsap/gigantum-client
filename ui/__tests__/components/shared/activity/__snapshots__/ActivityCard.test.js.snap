// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity renders a snapshot 1`] = `
<div
  className="ActivityCard__container column-1-span-9"
  data-selenium-id="ActivityCardx"
>
  <div
    className="ActivityCard ActivityCard--collapsed card ActivityCard--faded ActivityCard__expanded"
    style={Object {}}
  >
    <div
      className="ActivityCard__badge ActivityCard__badge--code"
      title="code"
    />
    <div
      className="ActivityCard__content"
    >
      <div
        className="ActivityCard__title flex flex--row justify--space-between closed"
        onClick={[Function]}
        role="presentation"
      >
        <div
          className="ActivityCard__stack"
        >
          <p
            className="ActivityCard__time"
          >
            3:10pm
          </p>
          <div
            className="ActivityCard__user"
          >
            C
          </div>
        </div>
        <h6
          className="ActivityCard__commit-message"
        >
          <b>
            cbutler - 
          </b>
          Created new Code directory \`code/folder\`
        </h6>
      </div>
    </div>
  </div>
</div>
`;
